drivers: gnss: add a generic NMEA GNSS driver

fabiobaltieri opened the pull request:
Minimal generic NMEA based UART GNSS driver, basically the Quectel driver without any specific stuff or power management. Prob worth adding PM at a later stage, a sleep pin + regulator or something, but let's start minimal for now.

Tested against an Antenova M10578-A3 and a random ebay ublox counterfeit module.

fabiobaltieri committed:
drivers: gnss: add a generic NMEA GNSS driver

Add a generic NMEA GNSS driver.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented on drivers/gnss/gnss_nmea_generic.c:
```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright 2023 Google LLC
```
should have prob left the original copyright since this is very the same quectel codebase

fabiobaltieri commented on drivers/gnss/gnss_nmea_generic.c:
```diff
@@ -0,0 +1,265 @@
+
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		ret = gnss_nmea_generic_suspend(dev);
```
going to put a "sleep" pin support here or something

fabiobaltieri commented on drivers/gnss/gnss_nmea_generic.c:
```diff
@@ -0,0 +1,265 @@
+		.satellites = data->satellites,
+		.satellites_size = ARRAY_SIZE(data->satellites),
+#endif
+		.timeout_ms = 200,
```
this defo needs to be a binding property

fabiobaltieri commented on drivers/gnss/gnss_nmea_generic.c:
```diff
@@ -0,0 +1,265 @@
+	DEVICE_DT_INST_DEFINE(inst, gnss_nmea_generic_init, PM_DEVICE_DT_INST_GET(inst),\
+			      &gnss_nmea_generic_data_##inst,				\
+			      &gnss_nmea_generic_cfg_##inst,				\
+			      POST_KERNEL, 99, &gnss_api);
```
`CONFIG_GNSS_INIT_PRIORITY`

fabiobaltieri commented on drivers/gnss/gnss_nmea_generic.c:
```diff
@@ -0,0 +1,265 @@
+		.uart = DEVICE_DT_GET(DT_INST_BUS(inst)),				\
+	};										\
+											\
+	static struct gnss_nmea_generic_data gnss_nmea_generic_data_##inst = {		\
+		.chat_delimiter = {'\r', '\n'},						\
+		.dynamic_separators_buf = {',', '*'},					\
+	};										\
```
this is wasting flash, should prob all be initialized in runtime

fabiobaltieri commented on drivers/gnss/gnss_nmea_generic.c:
```diff
@@ -0,0 +1,265 @@
+{
+	struct gnss_nmea_generic_data *data = dev->data;
+
+	data->dynamic_match.match = data->dynamic_match_buf;
+	data->dynamic_match.separators = data->dynamic_separators_buf;
+	data->dynamic_match.separators_size = sizeof(data->dynamic_separators_buf);
+	data->dynamic_match.wildcards = false;
+	data->dynamic_match.partial = false;
+
+	data->dynamic_script_chat.request = data->dynamic_request_buf;
+	data->dynamic_script_chat.response_matches = &data->dynamic_match;
+	data->dynamic_script_chat.response_matches_size = 1;
+	data->dynamic_script_chat.timeout = 0;
+
+	data->dynamic_script.name = "pair";
+	data->dynamic_script.script_chats = &data->dynamic_script_chat;
+	data->dynamic_script.script_chats_size = 1;
+	data->dynamic_script.abort_matches = NULL;
+	data->dynamic_script.abort_matches_size = 0;
+	data->dynamic_script.callback = NULL;
+	data->dynamic_script.timeout = 10;
```
these should prob go in flash

fabiobaltieri commented on drivers/gnss/gnss_nmea_generic.c:
```diff
@@ -0,0 +1,265 @@
+	struct k_spinlock lock;
+};
+
+#define MODEM_CHAT_SCRIPT_NO_ABORT_DEFINE(_sym, _script_chats, _callback, _timeout)             \
+	static struct modem_chat_script _sym = {                                                \
+		.name = #_sym,                                                                  \
+		.script_chats = _script_chats,                                                  \
+		.script_chats_size = ARRAY_SIZE(_script_chats),                                 \
+		.abort_matches = NULL,                                                          \
+		.abort_matches_size = 0,                                                        \
+		.callback = _callback,                                                          \
+		.timeout = _timeout,                                                            \
+	}
```
unused, drop

fabiobaltieri reviewed (commented):
there, selfie review, I would nack and request changes if I could

fabiobaltieri committed:
test: add a gnss build_all test

Add a generic test for GNSS devices.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Skimmed it down a bit more and tested again, added a build_all test as well. Should be good enough for a start.

cfriedt reviewed (commented):
Some nits, but would be magical even without magic numbers ✨

bjarki-trackunit commented on MAINTAINERS.yml:
```diff
@@ -1101,6 +1101,7 @@ Release Notes:
     - doc/hardware/peripherals/gnss.rst
     - drivers/gnss/
     - include/zephyr/drivers/gnss.h
+    - tests/drivers/build_all/input/
```
This should probably be tests/drivers/build_all/gnss/ :)

fabiobaltieri replied:
just when I thought I was on the top of my copy -> paste -> rename game, dammit.

bjarki-trackunit commented on tests/drivers/build_all/gnss/testcase.yaml:
```diff
@@ -0,0 +1,8 @@
+    - drivers
+    - gnss
+  build_only: true
+  platform_allow: native_posix
```
I believe we are moving to `native_sim` for all tests :)

bjarki-andreasen reviewed (changes requested):
Two small changes an we are good to go! I tested the driver on a Quectel L76LB as well, worked immediately :)

fabiobaltieri commented on drivers/gnss/gnss_nmea_generic.c:
```diff
@@ -0,0 +1,265 @@
+	/* Dynamic chat script */
+	uint8_t dynamic_match_buf[32];
+	uint8_t dynamic_separators_buf[2];
+	uint8_t dynamic_request_buf[32];
```
I don't even know what these are but they were not used so the magic is gone together with the non magic.

fabiobaltieri commented on drivers/gnss/gnss_nmea_generic.c:
```diff
@@ -0,0 +1,265 @@
+struct gnss_nmea_generic_data {
+	struct gnss_nmea0183_match_data match_data;
+#if CONFIG_GNSS_SATELLITES
+	struct gnss_satellite satellites[24];
```
Satellites are magical too, I mean seriously, how else would they work? Anyway, moved this to a Kconfig option so one can tune the amount of magic required.

fabiobaltieri replied:
✨

fabiobaltieri commented on drivers/gnss/gnss_nmea_generic.c:
```diff
@@ -0,0 +1,265 @@
+	struct modem_chat chat;
+	uint8_t chat_receive_buf[256];
+	uint8_t chat_delimiter[2];
+	uint8_t *chat_argv[32];
```
more defines

fabiobaltieri commented on drivers/gnss/gnss_nmea_generic.c:
```diff
@@ -0,0 +1,265 @@
+	/* UART backend */
+	struct modem_pipe *uart_pipe;
+	struct modem_backend_uart uart_backend;
+	uint8_t uart_backend_receive_buf[128];
```
twice as magic as 128, it's a define now

fabiobaltieri commented:
Update:
- moved few magics (✨) in defines and Kconfig
- fixed the native_sim and maintainer path
- also realized that few buffers were using `ARRAY_SIZE`, switched to `sizeof`, I think that would be correct one since that's used as bytes
- moved the delimiters on a file wide variable, this allowed dropping the initializer from the data structure which should save few bytes, could not make it const though, I think there's few fields in `struct modem_chat_config` that could use a `const` and then we could tweak some drivers to save some RAM

fabiobaltieri commented:
I tweaked the default `nmea-timeout-ms` up to 500, this is a sneaky one to get right, we could use a dedicated debugging option and some strategic debug print. Basically what I bumped into was that everything was working, then I disabled satellite support, that dropped the handler for GSV which sent the timing off and caused it to not work until I tweaked it up again.

bjarki-trackunit commented on dts/bindings/gnss/gnss-nmea-generic.yaml:
```diff
@@ -0,0 +1,32 @@
+      Synchronization timeout for NMEA sentences. The NMEA parser is expecting
+      to receive a GGA and RMC sentences within this time frame to publish a
+      location data. Set accordingly to the UART datarate and location
+      reporting frequency. Defaults to 200ms if unspecified.
```
500ms :)

fabiobaltieri replied:
Fixed. Do you think it would be possible to have the driver measure few messages and tweak this automatically?

bjarki-trackunit replied:
I think so, if we track the max delta between messages we are listening to, then we could set the timeout to half that delta.
```
0.10s RMC
0.15s GGA +0.05s
0.35s GSV +0.20s
0.36s GSV +0.01s
0.37s GSV +0.01s
0.38s GSV +0.01s
0.39s GSV +0.01s
1.00s RMC +0.61s
```
Max delta / 2 = 305ms
```
0.10s RMC
0.15s GGA +0.05s
0.35s GSV +0.20s
0.36s GSV +0.01s
0.37s GSV +0.01s
0.38s GSV +0.01s
0.39s GSV +0.01s
10.00s RMC +9.61s
```
Max delta / 2 = 4805ms

We can use RMC to recalculate the automatic delta calculation, to handle the user changing the fix rate. The first 1 to 2 data points would be discarded since the delta is not known

bjarki-trackunit replied:
Now that I'm looking at it, it can just be fix rate / 2 :D

fabiobaltieri replied:
but then if the link is slow enough (guess 4800 is still common on these) then a transmission I think could use well over 50% of the uart bandwidth and get close to the 1s fix. Maybe it could use like 90% fix rate and then reset once the message is published?

bjarki-trackunit replied:
The issue is the timing, we don't know which message comes first, RMC, GGA or GSV, if we don't get the timing right, we could be publishing an RMC message which contains data which does not reflect GGA or GSV. The timeout is for re-synchronization, so we know when to publish

bjarki-trackunit replied:
If we knew which came first, we could forego the timeout completely

fabiobaltieri replied:
Ok but then it sounds like the timeout option would only be usable on fast channels, if more than 50% of the uart time is utilized you'd end up never timing out and always reporting one of the two out of sync I imagine. But then the only solution would be to manually specify which one comes first... wondering how gps handles that.

bjarki-andreasen reviewed (commented):
In the Quectel driver, I could use proprietary NMEA commands to disable GSV and anything else I was not listening for to have a constant timeout, of course this can not be done for the generic one :)

One more nit though

cfriedt commented on drivers/gnss/gnss_nmea_generic.c:
```diff
@@ -0,0 +1,176 @@
+
+#define DT_DRV_COMPAT gnss_nmea_generic
+
+#define UART_RECV_BUF_SZ 128
+#define CHAT_RECV_BUF_SZ 256
+#define CHAT_ARGV_SZ 32
```
Could always convert to Kconfig, but not blocking
